src/KnetNLPModels.jl

```diff
@@ -99,6 +103,8 @@ function KnetNLPModel(
     x0,
     layers_g,
     nested_array,
+    1, #inti the batch current i to 1
```

Comment:
I don't understand the comment. Index of the first minibatch? Normally, you don't index iterators. You just loop through them.

Reply:
yes but this is because Knet doesn't use the iterator correctly 

The way knet does it is to ignore the size and do {i \in [1,data_size-batch_size]} which is terrible but that mean i for batch size of 1000 in MNIST is between 1 to 59999 which is weird for me but that is how it was written (I spent 2 days debugging till I find this weird behavior )


check here the knet random code: https://github.com/denizyuret/Knet.jl/blob/e9bec84b826ee7231db4ca8772fae6eb2fd4bcb9/src/train20/data.jl

function rand(d::Data)
i = rand(0:(d.length-d.batchsize))
return iterate(d, i)[1]
end

So basically their iterator is using i and i actually is the index of the data not the mini-batch, 
I am open to suggestion how to fix this since I think the problem is Knet's iterator is to some extend weird